src/chunktransformer.js

```diff
@@ -0,0 +1,337 @@
+/*!
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+const Mutation = require('./mutation.js');
+const stream = require('stream');
+const util = require('util');
+const Transform = stream.Transform;
+
+/**
+ * Enum for chunk formatter Row state.
+ * NEW_ROW: inital state or state after commitRow or resetRow
+ * ROW_IN_PROGRESS: state after first valid chunk without commitRow or resetRow
+ * CELL_IN_PROGRESS: state when valueSize > 0(partial cell)
+ */
+const RowStateEnum = Object.freeze({
+  NEW_ROW: 1,
+  ROW_IN_PROGRESS: 2,
+  CELL_IN_PROGRESS: 3,
+});
+/**
+ * ChunkTransformer formats all incoming chunks in to row
+ * keeps all intermediate state until end of stream.
+ * Should use new instance for each request.
+ */
+function ChunkTransformer(options) {
+  if (!(this instanceof ChunkTransformer)) {
+    return new ChunkTransformer(options);
+  }
+  this.options = options || {};
+  this.options.objectMode = true; // forcing object mode
+  Transform.call(this, options);
+  this._destroyed = false;
+  this.reset();
+}
+util.inherits(ChunkTransformer, Transform);
+/**
+ * transform the readrowsresponse chunks into friendly format. Chunks contain 3 properties:
+ *
+ * `rowContents` The row contents, this essentially is all data pertaining
+ *     to a single family.
+ *
+ * `commitRow` This is a boolean telling us the all previous chunks for this
+ *     row are ok to consume.
+ *
+ * `resetRow` This is a boolean telling us that all the previous chunks are to
+ *     be discarded.
+ *
+ * @public
+ *
+ * @param {object} data readrows response containing array of chunks.
+ * @param {object} [enc] encoding options.
+ * @param {callback} next callback will be called once data is processed, with error if any error in processing
+ */
+ChunkTransformer.prototype._transform = function(data, enc, next) {
+  try {
```

Comment:
Doing the `try/catch` inside each data event seems heavy, and like perf may be negatively affected. Is it possible that if the data is invalid, the called method, i.e. `processNewRow()`, could just destroy the stream right there with the error?

Reply:
@stephenplusplus even after calling destroy right there, current execution path from point of error to for loop needs to be short circuit. I see below option to achieve this. Let me know if there is any other way.


 Use _error instance object and it will be initialized in constructor as below:

```javascript
  this._error = {
    hasError: false,
    errorObj: null,
  };
```
Invariant will be 
```javascript
  if (condition) {
    this._error.hasError = true;
    this._error.errorObj = new TransformError({
      message: text,
      chunk: chunk,
    });
  }
```
processNewRow will be
```javascript
ChunkTransformer.prototype.processNewRow = function(chunk) {
  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);
  this.validateNewRow(chunk);
  if (this._error.hasError) {
    return;
  }
```
And transform will come out of for loop by checking on hasError
```javascript
ChunkTransformer.prototype._transform = function(data, enc, next) {
  for (const chunk of data.chunks) {
    if (this._error.hasError) {
      return next(this._error.errorObj);
    }
    switch (this.state) {
      case RowStateEnum.NEW_ROW:
        this.processNewRow(chunk, this.options);
        break;
      case RowStateEnum.ROW_IN_PROGRESS:
        this.processRowInProgress(chunk, this.options);
        break;
      case RowStateEnum.CELL_IN_PROGRESS:
        this.processCellInProgress(chunk, this.options);
        break;
    }
  }
  next();
};
```

In this case all conditions in validation method would still be checked only process* and transform/flush will be short circuited. If we make invariant return the condition we can use that to short circuit validation method as well like below.
```javascript
ChunkTransformer.prototype.invariant = function(condition, text, chunk) {
  if (condition) {
    this._error.hasError = true;
    this._error.errorObj = new TransformError({
      message: text,
      chunk: chunk,
    });
  }
  return condition;
};
```
validations method would be like this:
```javascript
ChunkTransformer.prototype.validateNewRow = function(chunk) {
  const row = this.row;
  const prevRowKey = this.prevRowKey;
  const newRowKey = Mutation.convertFromBytes(chunk.rowKey);
  return (
    this.invariant(
      typeof row.key !== 'undefined',
      'A new row cannot have existing state',
      chunk
    ) ||
    this.invariant(
      typeof chunk.rowKey === 'undefined' || chunk.rowKey === '',
      'A row key must be set',
      chunk
    ) ||
    this.invariant(chunk.resetRow, 'A new row cannot be reset', chunk) ||
    this.invariant(
      prevRowKey === newRowKey,
      'A commit happened but the same key followed',
      chunk
    ) ||
    this.invariant(!chunk.familyName, 'A family must be set', chunk) ||
    this.invariant(!chunk.qualifier, 'A column qualifier must be set', chunk) ||
    this.validateValueSizeAndCommitRow(chunk)
  );
};
```

if we make use of return value then we can use this return value all the way up to transform function to break for loop. 

My preference would be to use return value in case of error all the way up to transform and instance variable for error object. Let me know if you have any suggestion or better way of handling this.

Reply:
You should be able to re-use `this._destroyed` (set by `.destroy()`) to break out of the loop:

```js
ChunkTransformer.prototype._transform = function(data, enc, next) {
  for (const chunk of data.chunks) {
    if (this._destroyed) {
      return;
    }

    switch (this.state) {
      case RowStateEnum.NEW_ROW:
        this.processNewRow(chunk, this.options);
        break;
      case RowStateEnum.ROW_IN_PROGRESS:
        this.processRowInProgress(chunk, this.options);
        break;
      case RowStateEnum.CELL_IN_PROGRESS:
        this.processCellInProgress(chunk, this.options);
        break;
    }
  }

  next();
}
```

Reply:
Totally missed that. Thanks alot for pointing that out. I will update PR.